src/App.tsx

```diff
@@ -1,24 +1,222 @@
+  const [status, setStatus] = useState<Status>(Status.all);
+  const [todoTitle, setTodoTitle] = useState<string>('');
+  const [loadingTodoIds, setLoadingTodoIds] = useState<number[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+
+  const titleField = useRef<HTMLInputElement>(null);
+
+  const handleError = (errorMessage: string) => {
+    setError(errorMessage);
+    setTimeout(() => {
+      setError(null);
+    }, 3000);
+  };
+
+  useEffect(() => {
+    if (titleField.current) {
+      titleField.current.focus();
+    }
+  }, [todos]);
+
+  useEffect(() => {
+    todosService.getTodos(USER_ID)
+      .then(setTodos)
+      .catch(() => (handleError('Unable to load todos')));
+  }, []);
+
+  const visibleTodos = useMemo(() => {
+    return filteredTodos(todos, status);
+  }, [status, todos]);
+
+  const createTodo = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    const trimmedTitle = todoTitle.trim();
+
+    if (!trimmedTitle) {
+      handleError('Title should not be empty');
+    }
+
+    if (trimmedTitle) {
+      const tempTodoId = 0;
+
+      setTempTodo({
+        id: tempTodoId,
+        userId: USER_ID,
+        title: trimmedTitle,
+        completed: false,
+      });
+
+      setLoadingTodoIds(curr => [...curr, tempTodoId]);
+
+      todosService.addTodo({
+        userId: USER_ID, title: trimmedTitle, completed: false,
+      })
+        .then(newTodo => {
+          setTodos(currentTodos => {
+            return [...currentTodos, newTodo];
+          });
+          setTodoTitle('');
+        })
+        .catch(() => handleError('Unable to add a todo'))
+        .finally(() => {
+          setTempTodo(null);
+          setLoadingTodoIds(curr => curr.filter(id => id !== tempTodoId));
+        });
+    }
+  };
+
+  const deleteTodo = (todoId: number) => {
+    setLoadingTodoIds(curr => [...curr, todoId]);
+
+    todosService.deleteTodo(todoId)
+      .then(() => setTodos(
+        currentTodos => currentTodos.filter(todo => todo.id !== todoId),
+      ))
+      .catch(() => handleError('Unable to delete a todo'))
+      .finally(() => {
+        setLoadingTodoIds(curr => curr.filter(id => id !== todoId));
+      });
+  };
+
+  const clearTodos = () => {
+    todos.map(todo => {
+      if (todo.completed) {
+        deleteTodo(todo.id);
+      }
+
+      return 1;
+    });
+  };
+
+  const updateTodo = (newTodo: Todo): Promise<void> => {
+    setLoadingTodoIds(curr => [...curr, newTodo.id]);
+
+    return todosService.updateTodo(newTodo)
+      .then((todoFromServer) => setTodos(
+        currentTodos => currentTodos.map((todo) => {
+          const updatedTodo = todoFromServer as Todo;
+
+          if (todo.id === updatedTodo.id) {
+            return updatedTodo;
+          }
+
+          return todo;
+        }),
+      ))
+      .catch(() => {
+        handleError('Unable to update a todo');
+        throw new Error();
```

Comment:
There is no point in throwing an empty error, because it will not carry any information

Reply:
The thing is that without this throwing my handleEditSumbit function doesn't work correctly when it comes to showing an error message. To be more precise — the catch of this function doesn't work correctly when I'm trying to edit Todo in the offline mode.